btcclient/query.go

```diff
@@ -0,0 +1,199 @@
+	}
+
+	btcTxs := types.GetWrappedTxs(mBlock)
+	return types.NewIndexedBlock(int32(blockInfo.Height), &mBlock.Header, btcTxs), mBlock, nil
+}
+
+// GetBlockByHeight returns a block with the given height
+func (c *Client) GetBlockByHeight(height uint64) (*types.IndexedBlock, *wire.MsgBlock, error) {
+	blockHash, err := c.getBlockHashWithRetry(height)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get block by height %d: %w", height, err)
+	}
+
+	mBlock, err := c.getBlockWithRetry(blockHash)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get block by hash %s: %w", blockHash.String(), err)
+	}
+
+	btcTxs := types.GetWrappedTxs(mBlock)
+
+	return types.NewIndexedBlock(int32(height), &mBlock.Header, btcTxs), mBlock, nil
+}
+
+func (c *Client) getBestBlockHashWithRetry() (*chainhash.Hash, error) {
+	var (
+		blockHash *chainhash.Hash
+		err       error
+	)
+
+	if err := reporter.RetryDo(c.retrySleepTime, c.maxRetrySleepTime, func() error {
+		blockHash, err = c.GetBestBlockHash()
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		c.logger.Debug(
+			"failed to query the best block hash", zap.Error(err))
+	}
+
+	return blockHash, nil
+}
+
+func (c *Client) getBlockHashWithRetry(height uint64) (*chainhash.Hash, error) {
+	var (
+		blockHash *chainhash.Hash
+		err       error
+	)
+
+	if err := reporter.RetryDo(c.retrySleepTime, c.maxRetrySleepTime, func() error {
+		blockHash, err = c.GetBlockHash(int64(height))
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		c.logger.Debug(
+			"failed to query the block hash", zap.Uint64("height", height), zap.Error(err))
+	}
+
+	return blockHash, nil
+}
+
+func (c *Client) getBlockWithRetry(hash *chainhash.Hash) (*wire.MsgBlock, error) {
+	var (
+		block *wire.MsgBlock
+		err   error
+	)
+
+	if err := reporter.RetryDo(c.retrySleepTime, c.maxRetrySleepTime, func() error {
+		block, err = c.GetBlock(hash)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		c.logger.Debug(
+			"failed to query the block", zap.String("hash", hash.String()), zap.Error(err))
+	}
+
+	return block, nil
+}
+
+func (c *Client) getBlockVerboseWithRetry(hash *chainhash.Hash) (*btcjson.GetBlockVerboseResult, error) {
+	var (
+		blockVerbose *btcjson.GetBlockVerboseResult
+		err          error
+	)
+
+	if err := reporter.RetryDo(c.retrySleepTime, c.maxRetrySleepTime, func() error {
+		blockVerbose, err = c.GetBlockVerbose(hash)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		c.logger.Debug(
+			"failed to query the block verbose", zap.String("hash", hash.String()), zap.Error(err))
+	}
+
+	return blockVerbose, nil
+}
+
+// getChainBlocks returns a chain of indexed blocks from the block at baseHeight to the tipBlock
+// note: the caller needs to ensure that tipBlock is on the blockchain
+func (c *Client) getChainBlocks(baseHeight uint64, tipBlock *types.IndexedBlock) ([]*types.IndexedBlock, error) {
```

Comment:
I just want to confirm this is a wrapper of bitcoin RPC.

Reply:
Yes, this is a wrapper